File: Search_and_Sorting/java/BinarySearch.java
public class BinarySearch {

    public static int binarySearch(int[] arr,int val) {
        int start = 0, mid, end = arr.length - 1;
        while (start <= end) {
            mid = (start + end) / 2;
            if (arr[mid] == val) {
                return mid;
            } else if (arr[mid] < val) {
                start = mid + 1;
            } else {
                end = mid - 1;
            }
        }
        return -1;
    }
    public static void main(String[] args) {
        int[] arr = {1,2,3,4,5,6};
       int bs =  binarySearch(arr, 4);
        System.out.println(bs);
    }
}

File: Recursion_2/binarySearch.cpp
#include <iostream>
using namespace std;

// input - input array
// size - length of input array
// element - value to be searched
int binarySearchHelper(int input[],int start, int end, int element) {
    if (start <= end) {
        int mid = start + (end - start) / 2;
        if (input[mid] == element) { 
            return mid;
        }
        if (input[mid] > element) {
            return binarySearchHelper(input,start, mid - 1, element);
        }
         return binarySearchHelper(input,mid + 1,end,element);
    }
    return -1;
}
int binarySearch(int input[], int size, int element) {
    // Write your code here
    int start = 0;
    int end = size - 1;
    return binarySearchHelper(input,start,end,element);
}


int main() {
    int input[100000],length,element, ans;
    cin >> length;
    for(int i =0;i<length;i++)
    { 
        cin >> input[i];;
    }

    cin>>element;
    ans = binarySearch(input, length, element);
    cout<< ans << endl;
}
